code/modules/power/heavy_emitter.dm

```diff
@@ -0,0 +1,270 @@
+	firing = TRUE
+	icon_state = "centre"
+
+/obj/machinery/power/heavy_emitter/centre/turn_off()
+	if(!firing)
+		return
+	linked_interface.turn_off()
+	firing = FALSE
+	icon_state = "centre_off"
+
+/obj/machinery/power/heavy_emitter/centre/process(delta_time)
+	if(!firing || machine_stat & BROKEN)
+		return
+	if(surplus() < active_power_usage || !check_part_connectivity())
+		turn_off()
+		return
+
+	timer += delta_time
+
+	add_load(idle_power_usage)
+	if(timer >= max_timer)
+		timer = 0
+		add_load(active_power_usage)
+		radiation_pulse(src,500,can_contaminate=FALSE)
+		INVOKE_ASYNC(linked_cannon,/obj/machinery/power/heavy_emitter/cannon.proc/fire)
+		heat += 500
+
+	if(heat > max_heat)
+		explosion(src,4,8,16)
+		return
+
+	for(var/V in vents)
+		if(heat <= 0 || !V)
+			break
+		var/obj/machinery/power/heavy_emitter/vent/vent = V
+		if(vent.vent_gas())
+			heat -= 100
+
+
+/obj/machinery/power/heavy_emitter/arm
+	name = "Seismic Stabilizer Arm"
+	desc = "Reduces the knockback from firing to virtually nothing."
+	icon_state = "arm"
+
+/obj/machinery/power/heavy_emitter/interface
+	name = "Kinetic Amplification Manipulation Interface"
+	desc = "Allows for control over the core."
+	icon_state = "interface_off"
+	///Core connected to this thing
+	var/connected_core
+
+/obj/machinery/power/heavy_emitter/interface/attack_hand(mob/living/user)
+	. = ..()
+	if(connected_core)
+		var/obj/machinery/power/heavy_emitter/centre/centre = connected_core
+		if(!QDELETED(centre))
+			if(centre.firing)
+				centre.turn_off()
+			else
+				centre.turn_on()
+			return
+
+	var/turf/T = get_step(src,turn(dir,180))
+	var/obj/machinery/power/heavy_emitter/centre/centre = locate() in T
+	if(!centre)
+		turn_off()
+		return
+	if(centre.check_part_connectivity())
+		connected_core = centre
+		turn_on()
+	else
+		turn_off()
+
+/obj/machinery/power/heavy_emitter/interface/turn_on()
+	icon_state = "interface"
+
+/obj/machinery/power/heavy_emitter/interface/turn_off()
+	icon_state = "interface_off"
+	connected_core = null
+
+/obj/machinery/power/heavy_emitter/vent
+	name = "Energy Core Vent"
+	desc = "Circulates air around core preventing overheating"
+	icon_state = "vent"
+
+/obj/machinery/power/heavy_emitter/vent/proc/vent_gas()
+	var/turf/open/open_turf = get_step(src,dir)
+	if(!istype(open_turf))
+		return FALSE
+	var/datum/gas_mixture/gases = open_turf.return_air()
+	//Space can exist
+	if(gases)
+		gases.temperature += 100
+	flick("vent_on",src)
+	return TRUE
+
+/obj/machinery/power/heavy_emitter/cannon
+	name = "Energy Optic Converging Cannon"
+	desc = "Converges the energy of the core into singular destructive beam."
+	icon_state = "cannon"
+	var/warmup_sound = 'sound/machines/warmup1.ogg'
+	var/cooldown_sound = 'sound/machines/cooldown1.ogg'
+	var/projectile_sound = 'sound/weapons/beam_sniper.ogg'
+	var/projectile_type = /obj/projectile/beam/emitter/heavy
+
+/obj/machinery/power/heavy_emitter/cannon/proc/fire()
+	playsound(src, warmup_sound, 100)
+	sleep(5 SECONDS)
```

Comment:
i thought that we shouldn't use sleep() or there are some instances that is good to use them?

Reply:
you shouldnt use sleep in synchronous procs, this one is asynced so sleep is completely fine here.

Reply:
cool